File: leetcode/questions/505/python/505.py
import heapq

from typing import List

class Solution:
    def shortestDistance(self, maze: List[List[int]], start: List[int], destination: List[int]) -> int:
        INF = float('inf')
        steps = ((0, 1), (0, -1), (1, 0), (-1, 0))
        rows, cols = len(maze), len(maze[0])
        heap = [(0, start)]
        distances = [[INF for _ in range(cols)] for _ in range(rows)]
        distances[start[0]][start[1]] = 0

        while heap:
            dist, (r, c) = heapq.heappop(heap)
            
            if [r, c] == destination:
                break

            if dist > distances[r][c]:
                continue

            for r_inc, c_inc in steps:
                new_r, new_c = r + r_inc, c + c_inc
                walked = 0

                while 0 <= new_r < rows and 0 <= new_c < cols and maze[new_r][new_c] != 1:
                    new_r += r_inc
                    new_c += c_inc
                    walked += 1

                new_r -= r_inc
                new_c -= c_inc

                if distances[new_r][new_c] > dist + walked:
                    distances[new_r][new_c] = dist + walked
                    heapq.heappush(heap, (dist + walked, [new_r, new_c]))

        return distances[destination[0]][destination[1]] if distances[destination[0]][destination[1]] != INF else -1

File: leetcode/questions/505/java/505.java
import java.util.*;

class Solution {
    private int shortestPathDjikstra(int[] start, int[] end, int[][] maze) {
        int[][] steps = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        int rows = maze.length;
        int cols = maze[0].length;
        int [][] distances = new int[rows][cols];
        PriorityQueue<int []> heap = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        heap.add(new int[]{start[0], start[1], 0});

        for (int[] row: distances) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }

        distances[start[0]][start[1]] = 0;

        while (!heap.isEmpty()) {
            int[] node = heap.poll();
            int r = node[0], c = node[1], dist = node[2];

            if (dist > distances[r][c]) {
                continue;
            }

            if (end[0] == r && end[1] == c) {
                return dist;
            }

            for (int[] step: steps) {
                int newRow = r, newCol = c, numSteps = 0;

                do {
                    newRow += step[0];
                    newCol += step[1];
                    ++numSteps;
                } while (0 <= newRow && 0 <= newCol && newRow < rows && newCol < cols && maze[newRow][newCol] != 1);

                newRow -= step[0];
                newCol -= step[1];
                --numSteps; // Counted an extra step

                if (distances[newRow][newCol] > (dist + numSteps)) {
                    distances[newRow][newCol] = dist + numSteps;
                    heap.add(new int[] {newRow, newCol, distances[newRow][newCol]});
                }
            }
        }

        return -1;
    }

    public int shortestDistance(int[][] maze, int[] start, int[] destination) {
        return shortestPathDjikstra(start, destination, maze);
    }
}